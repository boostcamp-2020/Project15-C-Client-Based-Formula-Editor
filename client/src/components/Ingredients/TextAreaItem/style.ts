import styled from '@emotion/styled';

export const TextArea = styled.textarea`
  width: 100%;
  height: 100%;
  resize: none;
`;
