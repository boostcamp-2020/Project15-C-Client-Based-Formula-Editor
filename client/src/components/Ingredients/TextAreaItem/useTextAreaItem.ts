import { useDispatch } from 'react-redux';
import { editLatex } from '../../../contexts/latex';
import useCurrentTab from '@hooks/useCurrentTab';
import { useCallback, useState } from 'react';
import { FORMULA_HEADER, LatexContent } from '@constants/latex-header';
import { DISPLAY_INTERVAL } from '@constants/constants';
import useThrottle from '@hooks/useThrottle';
import useDebounce from '@hooks/useDebounce';

const useTextAreaItem = () => {
  const dispatch = useDispatch();
  const { currentTabInfo } = useCurrentTab();
  const [isShow, setIsShow] = useState(false);
  const [recommend, setRecommend] = useState<LatexContent[]>([]);
  const [maxNumber, setMaxNumber] = useState(DISPLAY_INTERVAL);

  const clearAndCloseRecommend = useCallback(() => {
    setIsShow(false);
    setRecommend([]);
  }, []);

  const onKeyPress = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    const target = e.target as HTMLTextAreaElement;
    if (e.key === 'Backspace' && target.value[target.value.length - 1] === '@') {
      return clearAndCloseRecommend();
    }

    if (e.key === 'Enter') return clearAndCloseRecommend();
    if (e.key === '@') {
      setIsShow(true);
    }
  };

  const recommendLatex = (targetWord: string) => {
    const startIndex = targetWord.lastIndexOf('@') + 1;
    const word = targetWord.substring(startIndex);
    const recommnedWords = FORMULA_HEADER.reduce<LatexContent[]>((acc, header) => {
      return acc.concat(header.content.filter((item) => item.latex.includes(word)));
    }, []);
    setRecommend(recommnedWords);
  };

  const debouncedRecommendLatex = useDebounce<[string]>(recommendLatex, 200);

  const onChangeHandler = (size: string) => (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    if (size === 'big' && isShow) {
      debouncedRecommendLatex(e.target.value);
    }
    dispatch(editLatex({ latex: e.target.value }));
  };

  const onScroll = (e: React.UIEvent<HTMLUListElement>) => {
    const target = e.target as HTMLUListElement;
    const scrollHeight = target.scrollHeight;
    const scrollTop = target.scrollTop;
    const clientHieght = target.clientHeight;

    if (scrollTop + clientHieght >= scrollHeight - 50 && maxNumber < recommend.length) {
      setMaxNumber((number) => number + DISPLAY_INTERVAL);
    }
  };

  const throttleOnScroll = useThrottle<[React.UIEvent<HTMLUListElement>]>(onScroll, 300);

  return {
    currentTabInfo,
    onChangeHandler,
    isShow,
    onKeyPress,
    recommend,
    clearAndCloseRecommend,
    maxNumber,
    onScroll: throttleOnScroll,
  };
};

export default useTextAreaItem;
