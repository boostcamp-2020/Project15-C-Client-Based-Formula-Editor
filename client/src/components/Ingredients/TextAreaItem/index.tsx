import RecommendItem from '@ingredients/RecommendItem';
import React from 'react';
import useTextAreaItem from './useTextAreaItem';
import * as S from './style';

interface TextAreaItem {
  size: string;
}

function TextAreaItem({ size }: TextAreaItem) {
  const {
    currentTabInfo,
    onChangeHandler,
    isShow,
    onKeyDown,
    onKeyUp,
    recommend,
    clearAndCloseRecommend,
    maxNumber,
    onScroll,
  } = useTextAreaItem();

  return (
    <S.TextAreaContainer>
      <S.TextArea
        value={currentTabInfo.latex}
        onChange={onChangeHandler(size)}
        onKeyDown={onKeyDown}
        onKeyUp={onKeyUp}
        isShow={isShow}
      />
      {size === 'big' && isShow && (
        <>
          {/* <S.Divider></S.Divider> */}
          <S.RecommendContainer>
            <S.Span>Tex Recommendation</S.Span>
            <ul onScroll={onScroll}>
              {recommend.slice(0, maxNumber).map((item, index) => (
                <RecommendItem
                  key={`${item.latex}${index}`}
                  currentLatex={currentTabInfo.latex}
                  clearAndCloseRecommend={clearAndCloseRecommend}
                  latex={item.latex}
                  image={item.image}
                />
              ))}
            </ul>
          </S.RecommendContainer>
        </>
      )}
    </S.TextAreaContainer>
  );
}

export default TextAreaItem;
