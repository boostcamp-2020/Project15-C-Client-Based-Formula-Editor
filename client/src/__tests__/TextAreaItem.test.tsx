import * as React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import TextAreaItem from '../components/Ingredients/TextAreaItem/index';
import '@testing-library/jest-dom';

describe('<TextAreaItem />', () => {
  it('renders TextAreaItem', () => {
    const { container } = render(<TextAreaItem size={'big'} />);
    expect(container).toMatchSnapshot();
  });
});
