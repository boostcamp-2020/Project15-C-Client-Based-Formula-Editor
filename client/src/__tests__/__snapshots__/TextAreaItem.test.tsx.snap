// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextAreaItem /> renders TextAreaItem 1`] = `
<div>
  <div
    class="css-1579fkq"
  >
    <textarea
      class="css-1kbq227"
    />
  </div>
</div>
`;
